std_symbol_map.inc

```diff
@@ -0,0 +1,2374 @@
+{ "std::logical_or<void>", kPrivate, "<functional>", kPublic },
+{ "std::logl", kPrivate, "<cmath>", kPublic },
+{ "std::lognormal_distribution", kPrivate, "<random>", kPublic },
+{ "std::longjmp", kPrivate, "<csetjmp>", kPublic },
+{ "std::lower_bound", kPrivate, "<algorithm>", kPublic },
+{ "std::lrint", kPrivate, "<cmath>", kPublic },
+{ "std::lrintf", kPrivate, "<cmath>", kPublic },
+{ "std::lrintl", kPrivate, "<cmath>", kPublic },
+{ "std::lround", kPrivate, "<cmath>", kPublic },
+{ "std::lroundf", kPrivate, "<cmath>", kPublic },
+{ "std::lroundl", kPrivate, "<cmath>", kPublic },
+{ "std::make_any", kPrivate, "<any>", kPublic },
+{ "std::make_const_iterator", kPrivate, "<iterator>", kPublic },
+{ "std::make_const_sentinel", kPrivate, "<iterator>", kPublic },
+{ "std::make_exception_ptr", kPrivate, "<exception>", kPublic },
+{ "std::make_format_args", kPrivate, "<format>", kPublic },
+{ "std::make_from_tuple", kPrivate, "<tuple>", kPublic },
+{ "std::make_heap", kPrivate, "<algorithm>", kPublic },
+{ "std::make_index_sequence", kPrivate, "<utility>", kPublic },
+{ "std::make_integer_sequence", kPrivate, "<utility>", kPublic },
+{ "std::make_move_iterator", kPrivate, "<iterator>", kPublic },
+{ "std::make_obj_using_allocator", kPrivate, "<memory>", kPublic },
+{ "std::make_optional", kPrivate, "<optional>", kPublic },
+{ "std::make_pair", kPrivate, "<utility>", kPublic },
+{ "std::make_reverse_iterator", kPrivate, "<iterator>", kPublic },
+{ "std::make_shared", kPrivate, "<memory>", kPublic },
+{ "std::make_shared_for_overwrite", kPrivate, "<memory>", kPublic },
+{ "std::make_signed", kPrivate, "<type_traits>", kPublic },
+{ "std::make_signed_t", kPrivate, "<type_traits>", kPublic },
+{ "std::make_tuple", kPrivate, "<tuple>", kPublic },
+{ "std::make_unique", kPrivate, "<memory>", kPublic },
+{ "std::make_unique_for_overwrite", kPrivate, "<memory>", kPublic },
+{ "std::make_unsigned", kPrivate, "<type_traits>", kPublic },
+{ "std::make_unsigned_t", kPrivate, "<type_traits>", kPublic },
+{ "std::make_wformat_args", kPrivate, "<format>", kPublic },
+{ "std::malloc", kPrivate, "<cstdlib>", kPublic },
+{ "std::map", kPrivate, "<map>", kPublic },
+{ "std::mask_array", kPrivate, "<valarray>", kPublic },
+{ "std::match_results", kPrivate, "<regex>", kPublic },
+{ "std::max", kPrivate, "<algorithm>", kPublic },
+{ "std::max_align_t", kPrivate, "<cstddef>", kPublic },
+{ "std::max_element", kPrivate, "<algorithm>", kPublic },
+{ "std::mblen", kPrivate, "<cstdlib>", kPublic },
+{ "std::mbrlen", kPrivate, "<cwchar>", kPublic },
+{ "std::mbrtoc16", kPrivate, "<cuchar>", kPublic },
+{ "std::mbrtoc32", kPrivate, "<cuchar>", kPublic },
+{ "std::mbrtoc8", kPrivate, "<cuchar>", kPublic },
+{ "std::mbrtowc", kPrivate, "<cwchar>", kPublic },
+{ "std::mbsinit", kPrivate, "<cwchar>", kPublic },
+{ "std::mbsrtowcs", kPrivate, "<cwchar>", kPublic },
+{ "std::mbstate_t", kPrivate, "<cuchar>", kPublic },
+{ "std::mbstate_t", kPrivate, "<cwchar>", kPublic },
```

Comment:
I'm not sure about this. Cppreference also [places](https://en.cppreference.com/w/cpp/string/multibyte/mbstate_t.html) `<cuchar>` before `<cwchar>` for `mbstate_t`. However, [it is noted](https://eel.is/c++draft/char.traits.specializations.char#1) in [char.traits.specializations.char] that

> The type `mbstate_t` is defined in [`<cwchar>`](https://eel.is/c++draft/cwchar.syn#header:%3ccwchar%3e)

So maybe, a canonical header list should be added to the script right now...

Reply:
But that mapping will essentially list all symbols, right? Or only the special ones where there's ambiguity? It seems like it's going to turn into quite a big hand-maintained thing.

Reply:
There are also `std::tuple_size` and `std::tuple_element`. I got stupid and thought that because the primary templates are not defined, IWYU should always suggest the correct header for a specialization. But actually, when a user _defines its own specialization_, it probably wants `<tuple>` to be suggested for his specialization header and not `<array>`.

Reply:
ISO C documents <wchar.h> as the canonical header for `wint_t`.

Reply:
```
$ man -w wint_t | xargs mansect 'SYNOPSIS|NOTES' | man /dev/stdin | cat
wint_t(3type)                                                     wint_t(3type)

SYNOPSIS
     #include <wchar.h>

     typedef /* ... */  wint_t;

     #define WEOF        /* ... */

     #include <stdint.h>

     #define WINT_WIDTH  /* ... */
     #define WINT_MAX    /* ... */
     #define WINT_MIN    /* ... */

NOTES
     The following header also provides wint_t and WEOF: <wctype.h>.

Linux man‐pages 6.16‐434‐gaf802... 2025‐12‐25                     wint_t(3type)
```

Reply:
> ISO C documents <wchar.h> as the canonical header for `wint_t`.

Maybe, you have made this conclusion just because `<wchar.h>` is documented before `<wctype.h>` in the standard? Anyway, in the proposed mapping, `<cwchar>` is placed before `<cwctype>`, according to the alphabetical order.

Reply:
The C++ C-compatibility headers are quite annoying because their behavior is/used to be poorly specified, and sometimes in conflict with the corresponding C headers. I wonder if maybe it's even a bad idea to consider the C standard for C++ headers? If we follow the letter of the C++ standard, at least we should be internally consistent.

Reply:
> So, for C identifiers, the C standard is indeed what should be followed, and the C++ standard should be taken with a pinch of salt.

I agree with this, but I'm still not fully confident that "described in ..." always means "the canonical header is ..."

If the C++ standard had any clear indication, I'd take that.

Reply:
I've taken a look at glibc v. 2.36, and it defines `wint_t` in a separate header `<bits/types/wint_t.h>`. Moreover, `<wctype.h>` doesn't include `<wchar.h>`, which is not so small. Hence, requiring a user to include `<wchar.h>` when he only uses `wint_t` and e.g. `towlower`, you'd actually make worse.

Reply:
I'd say that's more an exception than the rule.

Reply:
But I agree in some cases having the more lenient rules can be good, and thus I think it's make sense to have `--canonical-only` and default to not using it.

Reply:
I guess that makes sense. I don't have a clear idea of how to implement it, particularly since we don't have a strong concept of "canonical header". And I'm also not sure how it interplays with the current set-cover algorithm used to decide the desired headers. But it rings true!

Reply:
Seems we have an agreement.  :-)